packages/trpc/server/routers/loggedInViewer/outOfOffice.utils.ts

```diff
@@ -0,0 +1,41 @@
+import prisma from "@calcom/prisma";
+import { MembershipRole } from "@calcom/prisma/enums";
+
+export const isAdminForUser = async (adminUserId: number, memberUserId: number) => {
+  const user = await prisma.user.findUnique({
+    where: {
+      id: memberUserId,
+      teams: {
+        some: {
+          team: {
+            AND: [
+              {
+                members: {
+                  some: {
+                    userId: adminUserId,
+                    role: {
+                      in: [MembershipRole.ADMIN, MembershipRole.OWNER],
+                    },
+                    accepted: true,
+                  },
+                },
+              },
+              {
+                members: {
+                  some: {
+                    userId: memberUserId,
+                    accepted: true,
```

Comment:
This is very expensive. We already have a utility function to check if a user is an admin or owner

Reply:
I don't think this will be expensive, query is done on User (master) table using primary key. 
Can you please point me to the utility function, i could not find one that does the same (check input user is admin for the other user).

Reply:
Can you please provide either the execution plan or the raw SQL this produces? We need to make sure this won’t cause perf problems before merging.

Reply:
For some reason i interpreted `teams` column in `User` table as linking to `Team` table, but actually it is linked to 'Membership` table. May be i overlooked and went by the column name while building the query.
https://github.com/calcom/cal.com/blob/5bbccad41d625250ac03a6683e751b8ed6650721/packages/prisma/schema.prisma#L270

With this (knowing User table referencing to Membership table directly), i optimized the query further and came up with below query:

```
const user = await prisma.user.findUnique({
     where: {
      id: adminUserId,
      teams: {
        some: {
          accepted: true,
          role: {
            in: [MembershipRole.ADMIN, MembershipRole.OWNER],
          },
          team: {
            members: {
              some: {
                userId: memberUserId,
                accepted: true,
              },
            },
          },
        },
      },
    },
    select: {
      id: true,
    },
  });
```

For which prisma produces the raw SQL Query like below:
```
SELECT "public"."users"."id" 
FROM "public"."users" 
WHERE ("public"."users"."id" = $1 
AND ("public"."users"."id") IN (
    SELECT "t1"."userId" 
    FROM "public"."Membership" AS "t1" 
    LEFT JOIN "public"."Team" AS "j2" 
    ON ("j2"."id") = ("t1"."teamId") 
    WHERE ("t1"."accepted" = $2 
    AND (("j2"."id") IN (
        SELECT "t3"."teamId" 
        FROM "public"."Membership" AS "t3" 
        WHERE ("t3"."userId" = $3 
        AND "t3"."accepted" = $4 
        AND "t3"."role" IN (
            CAST($5::text AS "public"."MembershipRole"),
            CAST($6::text AS "public"."MembershipRole")
        ) 
        AND "t3"."teamId" IS NOT NULL
    )) 
    AND ("j2"."id" IS NOT NULL)) 
    AND "t1"."userId" IS NOT NULL
)) 
LIMIT $7 OFFSET $8;
```

Still the raw SQL Query contains join with Sub-query which is not efficient, Prisma some time produces the query with sub-queries though we intend to produce query with joins (which comparitively effecient than sub-queries).

So, Finally decided to come with 2 split simple queries that results raw SQL without joins.
```
const adminTeams = await prisma.membership.findMany({
    where: {
      userId: adminUserId,
      accepted: true,
      role: {
        in: [MembershipRole.ADMIN, MembershipRole.OWNER],
      },
    },
    select: {
      teamId: true,
    },
  });
  const adminTeamIds = adminTeams?.map((team) => team.teamId);
  const member = await prisma.membership.findFirst({
    where: {
      userId: memberUserId,
      accepted: true,
      teamId: {
        in: adminTeamIds,
      },
    },
    select: {
      id: true,
    },
  });
  ```
  The above prisma queries produces these raw SQL Queries
  
<img width="1507" alt="Screenshot 2024-09-26 at 5 57 56 PM" src="https://github.com/user-attachments/assets/39fdbc6f-288a-4572-9b7c-c52950129564">

since `userId` is indexed in `Membership` table, having these 2 simple queries, I hope should not be a performance issue.

Reply:
The final raw SQL Queries generated from `isAdminForUser()` function in `outOfOffice.utilt.ts`, in text format (same as snapshot in previous comment)

```
SELECT "public"."Membership"."id", "public"."Membership"."teamId" 
FROM "public"."Membership" 
WHERE ("public"."Membership"."userId" = $1 
       AND "public"."Membership"."accepted" = $2 
       AND "public"."Membership"."role" IN (CAST($3::text AS "public"."MembershipRole"), 
                                             CAST($4::text AS "public"."MembershipRole"))) 
OFFSET $5

SELECT "public"."Membership"."id" 
FROM "public"."Membership" 
WHERE ("public"."Membership"."userId" = $1 
       AND "public"."Membership"."accepted" = $2 
       AND "public"."Membership"."teamId" IN ($3, $4)) 
LIMIT $5 OFFSET $6
```

Reply:
@keithwillcode , i have attached the raw SQL queries generated from prisma query in 'isAdminForUser()' fn , can you please have look and let know if this is ok.